File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isUpper, isAlpha, isSpace)

responseFor :: String -> String
responseFor sentence
    | isSilence s = "Fine. Be that way!"
    | isShout s = "Whoa, chill out!"
    | isQuestion s = "Sure."
    | otherwise = "Whatever."
    where
        s = trim sentence
        isSilence = null
        isShout s = not (null alphas) && all isUpper alphas
            where
                alphas = filter isAlpha s
        isQuestion s = last s == '?'

trim :: String -> String
trim = ltrim . rtrim
    where
        ltrim = dropWhile isSpace
        rtrim = reverse . dropWhile isSpace . reverse

File: go/bob/bob.go
package bob

import (
	"regexp"
	"strings"
)

func Hey(remark string) string {
	s := strings.TrimSpace(remark)

	if s == "" {
		return "Fine. Be that way!"
	}

	uppers := regexp.MustCompile(`[A-Z]`)
	if uppers.MatchString(s) && strings.ToUpper(s) == s {
		return "Whoa, chill out!"
	}

	if strings.HasSuffix(s, "?") {
		return "Sure."
	}
	return "Whatever."
}
